src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -11,66 +11,194 @@
+    private static final String VALUE = "value";
+    private static final ClassLoader classLoader = ChatServlet.class.getClassLoader();
+
+    enum Attribute {
+        TOXICITY,
+        SEVERE_TOXICITY
+    }
+
+    /**
+     * When called, doGet will query all previously posted messages to the chat,
+     * and render the messages into a HTML template.
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String groupID = (String) request.getParameter(GROUP_ID_PARAM);
+
+        // This will be a placeholder until we can coordinate how to pass
+        // groupID from page to page
+        if (groupID == null) {
+            groupID = "123";
+        }
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        // This is just a sample group until group creation is coordinated
+        Entity sampleGroup = new Entity(GROUP_KIND, "BLM");
+        sampleGroup.setProperty(GROUP_NAME_PROPERTY, "Black Lives Matter");
+        datastore.put(sampleGroup);
+
+        // Calls query on all entities of type Message
+        Query messageQuery = new Query(MESSAGE_KIND + groupID).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
+        PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+
+        Query groupQuery = new Query(GROUP_KIND);
+        PreparedQuery preparedGroupQuery = datastore.prepare(groupQuery);
+
+        ImmutableMap<String, ImmutableList<String>> data = getTemplateData(preparedMessageQuery, preparedGroupQuery);
+
+        final String out = getOutputString("chatPage", data);
+
+        response.getWriter().println(out);
+    }
+
+    /**
+     * Is called by a button on the chat form. doPost takes the message and its
+     * attributes (such as timestamp and user) and stores them into Datastore.
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+        String groupID = (String) request.getParameter(GROUP_ID_PARAM);
+        final long timestamp = System.currentTimeMillis();
+
+        // This will be a placeholder until we can coordinate how to pass
+        // groupID from page to page
+        if (groupID == null) {
+            groupID = "123";
+        }
+
+        if (messageText.length() == 0) {
+            response.sendRedirect("/chat");
+            return;
+        }
+
+        // Reads API Key and HTTP referer from file 
+        File apiKeyFile = new File(classLoader.getResource("keys.txt").getFile());
+        Scanner scanner = new Scanner(apiKeyFile);
+        final String apiKey = scanner.nextLine();
+        final String referer = scanner.nextLine();
+        scanner.close();
+
+        final String perspectiveURL = 
+            "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + apiKey;
```

Comment:
Change the "https....key=" to a constant String. 
Then below you can just have `getCommentScore(URL_BASE + apiKey, referer, messageText)`

Reply:
should URL_BASE be a class variable or a local variable because I only use it in one method?